accepted/2021/architecture-targeting.md

```diff
@@ -0,0 +1,498 @@
+
+Today, we talk about self-contained and framework-dependent as the two [deployment types](https://docs.microsoft.com/dotnet/core/deploying/). This model is misleading for two different reasons: the set isn't complete, and it's oriented exclusively on deployment.
+
+The following are the full set of "build types" (I'm changing the term) that can be used to greater or lesser degrees across development, testing, and production scenarios.
+
+* Portable framework-dependent app
+* RID-specific framework-dependent app
+* RID-specific self-contained app
+
+Note: The first build type is the default today.
+
+This list makes these build types look like a progression, however it is not. It's just the cross-product of two characteristics (RID targeting and deployment model), with one variant removed ("portable self-contained app") because it isn't technically possible.
+
+Portable apps are the least well-defined app type, even though it's the current default. On one hand, portable apps are for advanced scenarios that require a single set of assets to be run in multiple production environments. That's not common.
+
+On the other hand, portable apps enable developers to reliably publish apps built on Windows to run on Linux, for example, without any special knowledge. In this case, the "multiple environments" are dev and production. For example, apps that depends on a local database, like [SQLite](https://www.nuget.org/packages/SQLite/) or [Microsoft.Data.SqlClient](https://www.nuget.org/packages/Microsoft.Data.SqlClient/) will run into this scenario, since local databases almost always contain RID-specific assets.
+
+With portable apps, the developer doesn't need to worry about the fact that these database packages include multiple RID implementations. With RID-specific apps, they would.
+
+However, there is more to the story. For example, the `SQLite` package contains (at the time of writing) seven RID-specific implementations (see the following list). They range from `800k` to `4MB`. That means a portable app that uses SQLite will contain all seven copies of the database engine even though it only ever relies on one, and need to accept the significant size increase that comes with that.
+
+* `linux-x64`
+* `osx-x64`
+* `win10-arm`
+* `win10-x64`
+* `win10-x86`
+* `win7-x64`
+* `win7-x86`
+
+See the appendix for an analysis of portable app size metrics.
+
+Portable apps are easier to deploy to production for PaaS scenarios, like Azure Websites. As soon as developers adopt containers, it makes more sense to adopt our [documented RID-specific container patterns](https://github.com/dotnet/dotnet-docker/tree/main/samples/aspnetapp), which results in optimized apps.
+
+The reason to keep portable apps as the default is that it is easier for non-container web developers that target a different operating system than the one they use locally for development. Portable apps are significantly worse for all other scenarios, particularly client.
```

Comment:
> The reason to keep portable apps as the default is that it is easier for non-container web developers that target a different operating system than the one they use locally for development.

Do we have any evidence that this workflow is used by developers?

Reply:
I don't, but I expect it happens a lot. For example, anyone who publishes directly from Visual Studio to a cloud may benefit from this. That said, I'm not arguing this point just conceding that this could be a legit reason.